### DISCUSSION TITLE
comparing the transmitted and received data to verify

### DISCUSSION THREAD

**Discussion by wajidali4** [micropython/micropython#13231] [Core Development]
how to do automatic verification whether the transmitted and the received data are identical. If both the transmitted and received data are stored in the form of text files how could this be achieved and for example For binary format please share your ideas and source if any one has done it before successfully i hope this topic will help to bring knowledge to the community.Thanks

**Comment by rkompass** [comment]
The problem you mention is usually solved using CRC checksums. There is much info out there about those (just google CRC). If you need Micropython code I can help.

**Reply by wajidali4** [reply]
can you help me with micropython code i am using ESPNOW protocol. Thanks for your feedback

**Reply by rkompass** [reply]
Here is example code (just the CRC part):
```py
import binascii

msg = b'The quick brown fox jumps over the lazy dog'
print('msg:', msg)
crc = binascii.crc32(msg)
sent = msg + crc.to_bytes(4, 'little')
print('sent:', sent)
print('. . . . . . .')
received = bytearray(sent)

# one bit change: (feel free to uncomment the following line)
# received[4] = received[4] ^ 0b00010000

print('received:', received)
check = binascii.crc32(received)
check_ok = 0x2144df1c    # we should always get this
print('check:', hex(check), 'OK' if check == check_ok else '!! Wrong !!')
received_msg = received[0:-4]
print('received_msg:', received_msg)
```

**Comment by peterhinch** [comment]
A key point about CRC's is that they work on a unidirectional link. You can send the data with the CRC. The receiver computes the CRC of the received data and compares it to the received CRC.
